File: lib/src/metadata.ts
import { GetObjectCommand, PutObjectCommand, type S3Client } from "@aws-sdk/client-s3";
import type { LadderRung } from "./config.js";
import { metadataKey } from "./contentId.js";
import { isNotFound } from "./s3.js";

export interface OutputMetadata {
  contentId: string;
  encoderVersion: string;
  encodedAt: string;
  source: {
    width: number;
    height: number;
    durationSeconds: number;
    bitrateKbps?: number;
  };
  ladder: LadderRung[];
}

export async function readMetadata(
  client: S3Client,
  bucket: string,
  contentId: string,
): Promise<OutputMetadata | null> {
  try {
    const res = await client.send(
      new GetObjectCommand({ Bucket: bucket, Key: metadataKey(contentId) }),
    );
    if (!res.Body) return null;
    return JSON.parse(await res.Body.transformToString()) as OutputMetadata;
  } catch (err) {
    if (isNotFound(err)) return null;
    throw err;
  }
}

export async function writeMetadata(
  client: S3Client,
  bucket: string,
  metadata: OutputMetadata,
): Promise<void> {
  await client.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: metadataKey(metadata.contentId),
      Body: JSON.stringify(metadata, null, 2),
      ContentType: "application/json",
    }),
  );
}

File: core/metadata.go
package core

import (
	"context"

	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// MetadataSource captures the probed source characteristics.
type MetadataSource struct {
	Width           int     `json:"width"`
	Height          int     `json:"height"`
	DurationSeconds float64 `json:"durationSeconds"`
	BitrateKbps     *int    `json:"bitrateKbps,omitempty"`
}

// OutputMetadata is written to by-id/<contentID>/metadata.json.
type OutputMetadata struct {
	ContentID      string         `json:"contentId"`
	EncoderVersion string         `json:"encoderVersion"`
	EncodedAt      string         `json:"encodedAt"`
	Source         MetadataSource `json:"source"`
	Ladder         []LadderRung   `json:"ladder"`
}

// ReadMetadata returns a content ID's metadata, or nil if none exists.
func ReadMetadata(ctx context.Context, client *s3.Client, bucket, contentID string) (*OutputMetadata, error) {
	return getJSONObject[OutputMetadata](ctx, client, bucket, MetadataKey(contentID))
}

// WriteMetadata writes a content ID's metadata.json.
func WriteMetadata(ctx context.Context, client *s3.Client, bucket string, m OutputMetadata) error {
	return putJSONObject(ctx, client, bucket, MetadataKey(m.ContentID), m)
}
